src/TemplateString.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+namespace ipl\Html;
+
+use Exception;
+
+/**
+ * Render {{#mustache}}Mustache{{/mustache}}-like string from {@link ValidHtml} element arguments
+ *
+ * # Example Usage
+ * ```
+ * $info = new TemplateString(
+ *     'Follow the %s for more information on %s {{#mustache}}Mustache{{/mustache}}',
+ *     [
+ *         new Link('doc/html', 'HTML documentation'),
+ *         Html::tag('strong', 'HTML elements'),
+ *        [
+ *        'mustache' => Html::tag('span')
+ *        ]
+ *     ]
+ * );
+ * ```
+ */
+class TemplateString extends FormattedString
+{
+    /** @var array */
+    protected $templateArgs = [];
+
+    protected $pos = 0;
+
+    protected $string;
+
+    protected $length;
+
+    protected $unclosed = [];
+
+    public function __construct($format, $args = null)
+    {
+        foreach ($args as $key => $val) {
+            if (is_array($val) && is_string(key($val))) {
+                unset($args[$key]);
+                $this->templateArgs += $val;
+            }
+        }
+
+        parent::__construct($format, $args);
+    }
+
+    /**
+     * @param null $for
+     * @return HtmlDocument
+     * @throws Exception in case of missing template argument or unbounded open or close templates
+     */
+    protected function parseTemplates($for = null)
+    {
+        $buffer = '';
+
+        while ($this->pos < $this->length) {
+            $char = $this->readChar();
+
+            if ($char === '{') {
+                $this->pos++;
+                $start = $this->readChar() === '#';
+                $this->pos -= 1;
+
+                $end = $this->readChar() === '/';
+
+                $templateKey = $this->readUntil('}');
+                // if the string following '{{#' is read up to the last character or (length - 1)th character
+                // then it is not a template
+                if ($this->pos >= $this->length - 1) {
+                    $strlen = strlen($templateKey);
+                    $this->pos -= ($strlen + 3);
+                    $buffer .= $this->readUntil();
```

Comment:
Can't you just do this?: (`$templateHandle` is what I imagine being the result of the single call to `readChar` above :wink: )

```suggestion
                    $buffer = '{{' . $templateHandle . $templateKey;
```